### PR TITLE
Azure Sql via CDK

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR adds support for creating Azure SQL Database resources via the CDK. Here is the usage:

```chsarp
builder.AddSqlServer("sql").AsAzureSqlDatabaseConstruct().AddDatabase("db");
```

This will produce the following Bicep output:

```bicep
targetScope = 'resourceGroup'

@description('')
param location string = resourceGroup().location

@description('')
param principalId string

@description('')
param principalName string


resource sqlServer_l5O9GRsSn 'Microsoft.Sql/servers@2022-08-01-preview' = {
  name: toLower(take(concat('sql', uniqueString(resourceGroup().id)), 24))
  location: location
  properties: {
    version: '12.0'
    minimalTlsVersion: '1.2'
    publicNetworkAccess: 'Enabled'
    administrators: {
      administratorType: 'ActiveDirectory'
      login: principalName
      sid: principalId
      tenantId: subscription().tenantId
      azureADOnlyAuthentication: true
    }
  }
}

resource sqlFirewallRule_Zjpl1WAet 'Microsoft.Sql/servers/firewallRules@2020-11-01-preview' = {
  parent: sqlServer_l5O9GRsSn
  name: 'AllowAllWindowsAzureIps'
  properties: {
    startIpAddress: '0.0.0.0'
    endIpAddress: '0.0.0.0'
  }
}

resource sqlDatabase_nUUkbFiVl 'Microsoft.Sql/servers/databases@2022-08-01-preview' = {
  parent: sqlServer_l5O9GRsSn
  name: 'sqldb'
  location: location
  properties: {
  }
}

output sqlServerFqdn string = sqlServer_l5O9GRsSn.properties.fullyQualifiedDomainName
```

And the following manifest output:

```json
{
  "resources": {
    "storagesku": {
      "type": "parameter.v0",
      "value": "{storagesku.inputs.value}",
      "inputs": {
        "value": {
          "type": "string"
        }
      }
    },
    "locationOverride": {
      "type": "parameter.v0",
      "value": "{locationOverride.inputs.value}",
      "inputs": {
        "value": {
          "type": "string"
        }
      }
    },
    "storage": {
      "type": "azure.bicep.v0",
      "path": "storage.module.bicep",
      "params": {
        "principalId": "",
        "principalType": "",
        "storagesku": "{storagesku.value}",
        "locationOverride": "{locationOverride.value}"
      }
    },
    "blobs": {
      "type": "value.v0",
      "connectionString": "{storage.outputs.blobEndpoint}"
    },
    "sql": {
      "type": "azure.bicep.v0",
      "connectionString": "Server=tcp:{sql.outputs.sqlServerFqdn},1433;Encrypt=True;Authentication=\u0022Active Directory Default\u0022",
      "path": "sql.module.bicep",
      "params": {
        "principalId": "",
        "principalName": ""
      }
    },
    "sqldb": {
      "type": "value.v0",
      "connectionString": "{sql.connectionString};Database=sqldb"
    },
    "api": {
      "type": "project.v0",
      "path": "../CdkSample.ApiService/CdkSample.ApiService.csproj",
      "env": {
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES": "true",
        "ConnectionStrings__blobs": "{blobs.connectionString}",
        "ConnectionStrings__sqldb": "{sqldb.connectionString}"
      },
      "bindings": {
        "http": {
          "scheme": "http",
          "protocol": "tcp",
          "transport": "http"
        },
        "https": {
          "scheme": "https",
          "protocol": "tcp",
          "transport": "http"
        }
      }
    }
  }
}
```

### CODE REVIEW

File: src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs [hunk 1]
```diff
@@ -65,6 +66,67 @@ public static IResourceBuilder<SqlServerServerResource> AsAzureSqlDatabase(this
...
+            var sqlFirewall = new SqlFirewallRule(construct);
+            sqlFirewall.AssignProperty(x => x.StartIPAddress, "'0.0.0.0'");
+            sqlFirewall.AssignProperty(x => x.EndIPAddress, "'255.255.255.255'");
+
+            List<SqlDatabase> sqlDatabases = new List<SqlDatabase>();
+            foreach (var databaseNames in builder.Resource.Databases)
+            {
+                var databaseName = databaseNames.Value;
+                var sqlDatabase = new SqlDatabase(construct, sqlServer, databaseName);
+                sqlDatabase.AssignProperty(x => x.Location, "location"); // HACK
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:97]: Why is this needed? Is it because this isn't a child resource of the server?

Comment (mitchdenny) [reply src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:97]: Arm requires it. It would be good to do automatically.

File: playground/cdk/CdkSample.AppHost/sql.module.bicep [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  properties: {
+    version: '12.0'
+    minimalTlsVersion: '1.2'
+    publicNetworkAccess: 'Enabled'
+    administrators: {
+      administratorType: 'ActiveDirectory'
+      principalType: principalType
+      login: principalName
+      sid: principalId
+      tenantId: subscription().tenantId
```
Comment (JoshLove-msft) [inline playground/cdk/CdkSample.AppHost/sql.module.bicep:28]: I assume this is equivalent to tenant().tenantId?

File: src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs (hunk 1 above)
Comment (JoshLove-msft) [reply src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:117]: Fixed in https://github.com/Azure/azure-sdk-for-net/pull/42341

File: src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs [hunk 3]
```diff
@@ -71,4 +72,107 @@ private static IResourceBuilder<AzureSqlServerResource> ConfigureDefaults(this I
...
+            sqlServer.AssignProperty(x => x.Administrators.AdministratorType, "'ActiveDirectory'");
+            sqlServer.AssignProperty(x => x.Administrators.IsAzureADOnlyAuthenticationEnabled, "true");
+            sqlServer.AssignParameter(x => x.Administrators.Sid, construct.PrincipalIdParameter);
+            sqlServer.AssignParameter(x => x.Administrators.Login, construct.PrincipalName);
+            sqlServer.AssignProperty(x => x.Administrators.TenantId, "subscription().tenantId");
+
+            // TODO: This specially named firewall rule allows access to this database from all other
+            //       Azure services. There is a problem here. The constructor of SqlFirewall does not
+            //       take a parent object so it infers which SQL instance we are using from the
+            //       construct. if we have two sql instances in the construct that would be a problem
```
Comment (JoshLove-msft) [inline src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:91]: ack - will add a parent param

Comment (JoshLove-msft) [reply src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:91]: https://github.com/Azure/azure-sdk-for-net/pull/42341/commits/44f5067f77547eda27632eb39a2a9b5f664733a6

File: playground/cdk/CdkSample.AppHost/sql.module.bicep (hunk 2 above)
Comment (mitchdenny) [reply playground/cdk/CdkSample.AppHost/sql.module.bicep:24]: Yes. We should use `tenant().tenantId`

File: src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs [hunk 4]
```diff
@@ -71,4 +72,96 @@ private static IResourceBuilder<AzureSqlServerResource> ConfigureDefaults(this I
...
+        {
+            var sqlServer = new SqlServer(construct, builder.Resource.Name);
+
+            sqlServer.AssignProperty(x => x.Administrators.AdministratorType, "'ActiveDirectory'");
+            sqlServer.AssignProperty(x => x.Administrators.IsAzureADOnlyAuthenticationEnabled, "true");
+            sqlServer.AssignParameter(x => x.Administrators.Sid, construct.PrincipalIdParameter);
+            sqlServer.AssignParameter(x => x.Administrators.Login, construct.PrincipalName);
+            sqlServer.AssignProperty(x => x.Administrators.TenantId, "subscription().tenantId");
+
+            var azureServicesFirewallRule = new SqlFirewallRule(construct, sqlServer, "AllowAllWindowsAzureIps");
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:88]:
```suggestion
            var azureServicesFirewallRule = new SqlFirewallRule(construct, sqlServer, "AllowAllAzureIps");
```

File: src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs [hunk 5]
```diff
@@ -71,4 +72,96 @@ private static IResourceBuilder<AzureSqlServerResource> ConfigureDefaults(this I
...
+        return builder;
+    }
+
+    /// <summary>
+    /// Configures SQL Server resource to be deployed as Azure SQL Database (server).
+    /// </summary>
+    /// <param name="builder">The builder for the SQL Server resource.</param>
+    /// <param name="configureResource">Callback to customize the Azure resources that will be provisioned in Azure.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<SqlServerServerResource> PublishAsAzureSqlDatabaseConstruct(this IResourceBuilder<SqlServerServerResource> builder, Action<ResourceModuleConstruct, SqlServer, IEnumerable<SqlDatabase>>? configureResource = null)
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:152]: Feels like we should expose the database list on a context object or something, or maybe not at all? Can you inspect these from the sql server or construct? Is the idea to pass any auto created resource to the callback?

Comment (mitchdenny) [reply src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:152]: So I think we'll probably end up revisiting this. I like the idea of having a context, but I don't like the idea of having a context type for every single `AddAzureX`/`PublishAsAzureX` method.

So my cheap solution for now is to just add arguments to the function for things that people might want to customize. The construct gives them access to everything, but the extra arguments are mostly for ergonomics.

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Extensions/AzureSqlExtensions.cs:152]: If yes, then each should have a context (what if we add more default resources in the future??).

File: playground/cdk/CdkSample.ApiService/Program.cs [hunk 6]
```diff
@@ -2,16 +2,29 @@
...
 
-app.MapGet("/", async (BlobServiceClient bsc) =>
+app.MapGet("/", async (BlobServiceClient bsc, SqlContext context) =>
+{
+    return new
+    {
+        blobFiles = await TestBlobStorageAsync(bsc),
+        sqlRows = await TestSqlServerAsync(context)
+    };
+});
```
Comment (davidfowl) [inline playground/cdk/CdkSample.ApiService/Program.cs:23]: This is gonna be the slowest endpoint ever.

**Review by davidfowl — APPROVED** [review]

File: playground/cdk/CdkSample.ApiService/Program.cs (hunk 6 above)
Comment (mitchdenny) [reply playground/cdk/CdkSample.ApiService/Program.cs:23]: Having an API do something with SQL and blobs is pretty common ;)
